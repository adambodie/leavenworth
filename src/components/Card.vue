<template>
	<div>
		<div class="options">
			<div class="type">
			<select v-on:change="filterListByType">
				<option value="">Select type of cards...</option>
				<option v-for="(item, index) in uniqueCardsList" :key="index" v-bind:data-testid="item">{{item}}</option>
			</select>
			</div>
			<div class="filter">
				<select v-on:change="filterListByAmount">
					<option value="">Filter by Amount...</option>
					<option v-for="(card, index) in cards" v-bind:value="index" :key="index" v-bind:data-testid="index">
						{{ index + 1 }}
					</option>
				</select>
			</div>
		</div>
		<transition-group name="flip" tag="ul" class="card-list">
			<li 
				v-bind:class="{ invisible: !card.visible }"
				v-bind:data-testid="index"
				v-show="(index <= value && type === '') || type === card.type" 
				v-for="(card, index) in shuffleCards" 
				:key="card.id"
			>
				<vue-flip :active-click="true" width="259px" height="345px" transition="0.4s" >
					<p slot="front" class="card" key="english">{{card.english}}
						<span class="language">English</span>
						<span class="delete-card" v-on:click="removeCard(index)">X</span>
					</p>
					<p slot="back" class="card" key="german">{{card.german}}
						<span class="language">German</span>
						<span class="delete-card" v-on:click="removeCard(index)">X</span>
					</p>
				</vue-flip>
			</li>
		</transition-group>
	</div>
</template>

<script>
import VueFlip from 'vue-flip';
import { mapState } from 'vuex';

export default {
	components: {
		'vue-flip': VueFlip
	},
	methods: {
		filterListByType(){
			this.$store.dispatch('filterListByType');
		},
		filterListByAmount(){
			this.$store.dispatch('filterListByAmount');
		},
		removeCard(card) {
			this.$store.commit('removeCard', card);
		}
	},
	computed: mapState({
		cards: state => state.cards,
		value: state => state.value,
		type: state => state.type,
		visible: state => state.visible,
		uniqueCardsList() {
			const types = [];
			this.cards.forEach((card) => {
				if(!types.includes(card.type)) {
					types.push(card.type);
				}
			});
			return types;
		},
		shuffleCards() {
			const array = this.cards;
			let currentIndex = array.length, temporaryValue, randomIndex;
			while (0 !== currentIndex) {
				randomIndex = Math.floor(Math.random() * currentIndex);
				currentIndex -= 1;
				temporaryValue = array[currentIndex];
				array[currentIndex] = array[randomIndex];
				array[randomIndex] = temporaryValue;
			}
			return array;
		}
	})
}
</script>
