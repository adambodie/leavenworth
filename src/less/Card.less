//Imports

@import (css) url('https://fonts.googleapis.com/css?family=Shrikhand');

// Variables
@images: 'http://leavenworth.bodiewebdesign.com/images';
@font-family: 'Shrikhand', cursive;
@banner-color: #c79a61;

// Mixins
.card-image-size() {
	width: 225px;
	height: 300px;
}

.title-font-shadow() {
		font-family: @font-family;
		color: white;
		text-shadow: 2px 2px 4px black;
}

//Main Styles

body {
    font-family: 'Open Sans', sans-serif; 
    background-color: #a86957;
}

.border {
	position: absolute;
	left: 0;
	right: 0;
	width: 100%;
}

// Header Styles
#header {
	padding: 10px;
	background-color: @banner-color;
	h1 {
		font-size: 2.5em;
		font-family: @font-family;
		color: white;
		text-shadow: 4px 4px 8px darken(white, 80%);
	}
}

// Intro Styles
.intro-body {
	display: flex;
}

.paragraph {
	padding: 20px;
}

.map {
	img {
		padding: 20px;
	}
}

// Footer Styles
#footer {
	padding: 10px;
	background-color: @banner-color;
	font-family: @font-family;
	color: white;
}


// Jumbotron Styles
.jumbotron {
	margin: 30px;
	box-shadow: 4px 4px 8px #23201b;
	h1 {
		.title-font-shadow;
		text-align: center;
		font-size: 2em;
		padding: 20px;
		background-color: @banner-color;
		box-shadow: 4px 4px 8px #23201b;
	}
}

// Loading Background Styles
.loading-background {
    background: #a86957!important;
    opacity: 1!important;
}


// Photo Gallery Styles
.caption {
	position: absolute;
	left: 0;
	bottom: 10px;
	padding: 10px 10px;
	font-size: 14px;
	.title-font-shadow;
	text-align: center;
	width: 100%;
}

.owl-carousel .owl-item img {
    display: block;
    width: auto;
    height: 450px!important;
    margin: 0 auto;
}

.owl-item {
	img {
		opacity: 0.5;
	}
	.caption {
		display: none;
	}
}

.owl-nav {
	margin-top: 0;
}
.center {
	img {
		opacity: 1;
	}
	.caption {
		display: block;
	}
}

.owl-theme .owl-nav [class*=owl-] {
    font-size: 36px!important;
    margin: 20px;
    background: @banner-color!important;
    width: 50px;
    height: 50px;
    border-radius: 100%;
    box-shadow: 2px 2px 4px black;
    &:focus {
		outline: none;
    }
}

// Cards Styles
.options {
	display: flex;
	justify-content: space-around;
	align-items: center;
}
ul {
    padding-left: 0;
    display: flex;
    flex-flow: wrap;
    justify-content: center;
  }

li {
	list-style-type: none;
    padding: 10px 10px;
    transition: all 0.5s ease;
		&:nth-child(3n + 0) .card {
			background: url('@{images}/card1.JPG');
		}
		&:nth-child(3n + 1) .card {
			background: url('@{images}/card2.JPG');
		}
		&:nth-child(3n + 2) .card {
			background: url('@{images}/card3.JPG');
		}
	}


.card {
    display: flex;
    justify-content: center;
    align-items: center;
    .card-image-size;
    .title-font-shadow;
    border-radius: 20px;
    text-align: center;
    position: relative;
    font-size: 2em;
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
    will-change: transform;
  }

.delete-card {
	position: absolute;
	right: 0;
	top: 0;
	padding: 10px 15px;
	opacity: 0.4;
	transition: all 0.5s ease;
	font-family: 'Open Sans', sans-serif;
	font-size: 14px;
	cursor: pointer;
	&:hover {
		color: black;
	}
}
.language {
	position: absolute;
	left: 0;
	top: 0;
	padding: 10px 15px;
	transition: all 0.5s ease;
	font-size: 14px;
}

.invisible {
	display: none;
}

// Transition Styles
.flip-enter-active {
  transition: all 0.4s ease;
}
  
.flip-leave-active {
  display: none;
}
  
.flip-enter, .flip-leave {
  transform: rotateY(180deg);
  opacity: 0;
}