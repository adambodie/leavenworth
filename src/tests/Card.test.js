import Vue from 'vue'
import Card from '../vue/card.vue'
import { shallow } from 'vue-test-utils'

describe('Card.test.js', () => {
	let cmp, vm
	
	beforeEach(() => {
		cmp = shallow(Card, {})
		cmp.setData({cards: [
			{english: 'Beer', german: 'Bier', type: 'Cuisine', flipped: false, visible: true},
			{english: 'Mountain', german: 'Berg', type: 'Geography', flipped: false, visible: true},
			{english: 'Goat', german: 'Ziege', type: 'Animal', flipped: false, visible: true}
		], 
		checked: false, type: '', visible: true})
	})
	
	it('has the expected html structure', () => {
		expect(cmp.element).toMatchSnapshot()
	})
	it('should have a length of 3', () => {
		cmp.vm.types = [];
		expect(cmp.vm.uniqueCardsList).toContain('Animal')
		expect(cmp.vm.uniqueCardsList.length).toBe(3)
	})
})
