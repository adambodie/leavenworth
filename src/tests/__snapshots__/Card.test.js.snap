// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.test.js has the expected html structure 1`] = `
<div
  id="card"
>
  <h1>
    English-German Flash Cards
  </h1>
  <select>
    <option
      value=""
    >
      Select type of cards...
    </option>
    <option>
      Animal
    </option>
    <option>
      Cuisine
    </option>
    <option>
      Geography
    </option>
  </select>
  <div
    class="showAll"
  >
    <input
      type="checkbox"
    />
     Show All Cards
  </div>
  <ul
    class="card-list"
  >
    <li
      class=""
    >
      <p
        class="card"
        name="flip"
      >
        Mountain
        <span
          class="language"
        >
          English
        </span>
        <span
          class="delete-card"
        >
          X
        </span>
      </p>
    </li>
    <li
      class=""
    >
      <p
        class="card"
        name="flip"
      >
        Beer
        <span
          class="language"
        >
          English
        </span>
        <span
          class="delete-card"
        >
          X
        </span>
      </p>
    </li>
    <li
      class=""
    >
      <p
        class="card"
        name="flip"
      >
        Goat
        <span
          class="language"
        >
          English
        </span>
        <span
          class="delete-card"
        >
          X
        </span>
      </p>
    </li>
  </ul>
</div>
`;
